### ISSUE TITLE
Different behaviour for recursive TypedDict defined as class/functional

### ISSUE THREAD

**Issue by m0rl** [python/mypy#14460]
**Bug Report**

`TypedDict`-based recursive type is flagged as invalid when defined using functional syntax while no issue is reported for analogous class-based definition.

**To Reproduce**

```python
from typing import List, TypedDict

Example = TypedDict("Example", {"rec": List["Example"]})
```

**Expected Behavior**

`TypedDict`-based recursive types to have similar behaviour when defined using class/functional syntax.

**Actual Behavior**

```python
from typing import List, TypedDict

Example = TypedDict("Example", {"rec": List["Example"]})
```

results in an error:
> Cannot resolve name "Example" (possible cyclic definition)

while similar type defined as a class has no issues reported:

```python
from typing import List, TypedDict

class Example(TypedDict):
    rec: List["Example"]
```

Also no issue is reported when the recursive property is defined using some of the special forms (like `Required`, `Optional` etc) instead of `List`.  The following code passes typecheck without errors:

```python
from typing import Optional, TypedDict

Example = TypedDict("Example", {"rec": Optional["Example"]})
```

**Your Environment**

- Mypy version used: `mypy 0.991 (compiled: yes)`
- Mypy command-line flags: `none`
- Mypy configuration options from `mypy.ini` (and other config files): `none`
- Python version used: `3.11.1`
